lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -384,6 +384,259 @@ class DecomposeAtenReshapeOp : public OpRewritePattern<AtenReshapeOp> {
+    }
+
+
+    auto collectOperandDims =
+        [resultTokens](
+            SmallVector<Value> operandShape, SmallVector<char> operandTokens,
+            SmallVector<char> others, SmallVectorImpl<int64_t> &contractingDims,
+            SmallVectorImpl<int64_t> &batchingDims,
+            SmallVector<char> &dotResultTokens,
+            SmallVector<Value> &dotResultShape) {
+          llvm::SmallDenseSet<char> othersSet(others.begin(), others.end());
+          llvm::SmallDenseSet<char> resultTokensSet(resultTokens.begin(),
+                                                    resultTokens.end());
+          for (const auto &en : llvm::enumerate(operandTokens)) {
+            bool isResultToken = resultTokensSet.contains(en.value());
+            bool isOtherToken = othersSet.contains(en.value());
+            if (!isResultToken && isOtherToken) {
+              contractingDims.push_back(en.index());
+            } else if (isOtherToken) {
+              batchingDims.push_back(en.index());
+            } else {
+              dotResultTokens.push_back(en.value());
+              dotResultShape.push_back(operandShape[en.index()]);
+            }
+          }
+        };
+
+    for (int i = 0; i < totalRank - 1; i++) {
+      SmallVector<int64_t> lhsContractingDims, lhsBatchingDims,
+          rhsContractingDims, rhsBatchingDims;
+      SmallVector<char> dotResultTokens;
+      SmallVector<Value> dotResultShape;
+      SmallVector<Value> lhsShape = inputShapes[0];
+      SmallVector<Value> rhsShape = inputShapes[1];
+      SmallVector<char> lhsTokens, rhsTokens;
+      lhsTokens = inputTokens[0];
+      rhsTokens = inputTokens[1];
+      collectOperandDims(lhsShape, lhsTokens, rhsTokens, lhsContractingDims,
+                        lhsBatchingDims, dotResultTokens, dotResultShape);
+      collectOperandDims(rhsShape, rhsTokens, lhsTokens, rhsContractingDims,
+                        rhsBatchingDims, dotResultTokens, dotResultShape);
+      // Prepend batch tokens.
+      for (const auto &it : llvm::enumerate(lhsBatchingDims)) {
+        char batchingToken = lhsTokens[it.value()];
+        Value batchingShapeDim = lhsShape[it.value()];
+        dotResultTokens.insert(dotResultTokens.begin() + it.index(),
+                              batchingToken);
+        dotResultShape.insert(dotResultShape.begin() + it.index(),
+                              batchingShapeDim);
+      }
+      // Lowering to dot_general does not support a mismatch between the number
+      // of result dims and the number of non-contracting dims.
+      Value constZero =
+          rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(0));
+      Value constOne =
+          rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(1));
+      Value constTwo =
+          rewriter.create<ConstantIntOp>(loc, rewriter.getI64IntegerAttr(2));
+      SmallVector<Value> lhsFinalShape, rhsFinalShape, finalShape;
+      if (!lhsContractingDims.empty()) {
```

Comment:
All these `if-else`s seem to be a little similar. They all do some creation of ints, a transpose, and a swap of values. Is there a common logic that can be extracted to avoid repetition?